### PR TITLE
fix exception when non-data properties present

### PR DESCRIPTION
**Pull request by NathanMagnus**
An exception was thrown when a ValigatorModel (either ValigatorModel or ValigatorModelAttribute) had properties that were not of type `Data<>`.

The exception is an index out of range because it was trying to the generic on a property that isn't a `Data<>`. This PR fixes that by defining a separate code path at the point of serialization/deserialization for those properties that are not `Data<>`

### CODE REVIEW

File: Valigator.Core/Core/Model.cs [hunk 1]
```diff
@@ -241,9 +241,6 @@ private static IEnumerable<PropertyInfo> GetExplicitProperties(Type type)
 			}
 		}
 
-		private static bool IsValigatorDataType(Type type)
```
Comment (NathanMagnus) [inline Valigator.Core/Core/Model.cs:244]: Moved to a public extension method to use elsewhere too.

File: Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs [hunk 2]
```diff
@@ -17,12 +17,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
...
+		public NonValigatorDataJsonPropertyHandler(Func<TObject, TValue> getValue, Action<TObject, TValue> setValue)
+		{
+			_getValue = getValue;
+			_setValue = setValue;
+		}
+
+		public override void ReadProperty(JsonReader reader, JsonSerializer serializer, TObject obj)
+		{
+			if (_setValue != null)
+				_setValue(obj, serializer.Deserialize<TValue>(reader));
```
Comment (NathanMagnus) [inline Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs:74]: In the case where it isn't a `Data<>` just use the serializer as usual to deserialize, but we need to set using the _setValue.

File: Valigator.Core/Valigator.Core.csproj [hunk 3]
```diff
@@ -13,8 +13,4 @@
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
   </ItemGroup>
 
-  <ItemGroup>
-    <Folder Include="Extensions\" />
```
Comment (michaelr-iq) [inline Valigator.Core/Valigator.Core.csproj:17]: Was the folder removed on purpose?

File: Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs [hunk 4]
```diff
@@ -17,12 +17,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
...
+			else
+				throw new NotSupportedException();
+		}
+
+		public override void WriteProperty(JsonWriter writer, JsonSerializer serializer, TObject obj)
+		{
+			if (_getValue != null)
+				serializer.Serialize(writer, _getValue(obj), typeof(TValue));
+			else
+				throw new NotSupportedException();
```
Comment (michaelr-iq) [inline Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs:84]: Can you provide a more detailed string to the NotSupported exception?

File: Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs [hunk 5]
```diff
@@ -17,12 +17,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
...
+		}
+
+		public new static NonValigatorDataJsonPropertyHandler<TObject, TValue> Create(PropertyInfo property)
+		{
+			if (property.GetCustomAttribute<JsonIgnoreAttribute>() != null)
+				return new NonValigatorDataJsonPropertyHandler<TObject, TValue>(null, null);
+
```
Comment (michaelr-iq) [inline Valigator.Newtonsoft.Json/ValigatorJsonPropertyHandler.cs:91]: double newline in case you touch the file

File: Valigator.TestApi.Core31/Controllers/TestController.cs [hunk 6]
```diff
@@ -26,7 +26,24 @@ public JsonResult Post([FromBody] GuidBodyClass bodyValue)
 
 		[HttpPost("anonymousObjectOutput")]
 		public JsonResult PostReturnsAnonymousObject([FromBody] GuidBodyClass bodyValue)
-			=> new JsonResult(new { TheOutputOfThePost = bodyValue.IdentifierCollection }.ToValigatorModel());
+			=> new JsonResult(new { TheOutputOfThePost = bodyValue.IdentifierCollection, SecondValue = 1 }.ToValigatorModel());
+
+		[HttpPost("nonDataPropertyObject")]
+		public JsonResult PostWithNonDataPropertyObject([FromBody] GuidBodyClass bodyValue)
+		=> new JsonResult(new NonDataPropertyModel(bodyValue.IdentifierCollection.Value));
```
Comment (michaelr-iq) [inline Valigator.TestApi.Core31/Controllers/TestController.cs:33]: tab

File: Valigator.Text.Json/ValigatorJsonPropertyHandler.cs [hunk 7]
```diff
@@ -18,12 +18,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
 
 		public static ValigatorJsonPropertyHandler<TObject> Create(PropertyInfo property)
 		{
-			var handlerType = typeof(ValigatorJsonPropertyHandler<,>).MakeGenericType(typeof(TObject), property.PropertyType.GetGenericArguments()[0]);
+			var propertyHandlerType = property.PropertyType.IsValigatorDataType() ? typeof(ValigatorJsonPropertyHandler<,>) : typeof(NonValigatorDataJsonPropertyHandler<,>);
+
+			var propertyType = property.PropertyType.IsValigatorDataType() ? property.PropertyType.GetGenericArguments()[0] : property.PropertyType;
```
Comment (michaelr-iq) [inline Valigator.Text.Json/ValigatorJsonPropertyHandler.cs:23]: can dry this out by extracting the logic into two seperate methods that return a create method depending on IsValigatorDataType() instead of checking twice for IsValigatorDataType(

File: Valigator.Text.Json/ValigatorJsonPropertyHandler.cs [hunk 8]
```diff
@@ -18,12 +18,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
...
 		}
+
+		internal static Func<TObject, TValue> CreateGetter<TValue>(PropertyInfo property)
+		{
+			var objParameter = Expression.Parameter(typeof(TObject), "obj");
+			var propertyParameter = Expression.Constant(property);
+
+			var getValueMethod = property.GetType().GetMethod(nameof(PropertyInfo.GetValue), new[] { typeof(object) });
+			var getterExpression = Expression.Convert(Expression.Call(propertyParameter, getValueMethod, objParameter), property.PropertyType);
+			var getter = property.CanRead ? Expression.Lambda<Func<TObject, TValue>>(getterExpression, objParameter).Compile() : null;
```
Comment (michaelr-iq) [inline Valigator.Text.Json/ValigatorJsonPropertyHandler.cs:39]: Is there a way to avoid the null? or prvoide a lambda that throws a specific exception?

File: Valigator.Text.Json/ValigatorJsonPropertyHandler.cs [hunk 9]
```diff
@@ -18,12 +18,84 @@ internal abstract class ValigatorJsonPropertyHandler<TObject>
...
+
+		internal static Action<TObject, TValue> CreateSetter<TValue>(PropertyInfo property)
+		{
+			var objParameter = Expression.Parameter(typeof(TObject), "obj");
+			var propertyParameter = Expression.Constant(property);
+
+			var valueParameter = Expression.Parameter(typeof(TValue), "value");
+			var setValueMethod = property.GetType().GetMethod(nameof(PropertyInfo.SetValue), new[] { typeof(object), typeof(object) });
+			var setProperty = Expression.Call(propertyParameter, setValueMethod, objParameter, Expression.Convert(valueParameter, typeof(object)));
+			var setter = property.CanWrite ? Expression.Lambda<Action<TObject, TValue>>(setProperty, objParameter, valueParameter).Compile() : null;
```
Comment (michaelr-iq) [inline Valigator.Text.Json/ValigatorJsonPropertyHandler.cs:52]: Is there a way to avoid the null? or prvoide a lambda that throws a specific exception?

Comment (michaelr-iq) [reply Valigator.Text.Json/ValigatorJsonPropertyHandler.cs:52]: I guess you check canWRite and can read later on.

File: Valigator.sln [hunk 10]
```diff
@@ -35,6 +35,11 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Valigator.Tests.Common", "V
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{FA1E881C-D1A5-4595-B52A-2113A91B2EF1}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{7EBB667E-7531-4D69-A828-B326690267B6}"
```
Comment (michaelr-iq) [inline Valigator.sln:38]: Is this change intended?

**Review by michaelr-iq — CHANGES_REQUESTED** [review]

**Review by michaelr-iq — APPROVED** [review]

File: Valigator.sln (hunk 10 above)
Comment (NathanMagnus) [reply Valigator.sln:38]: Yes

File: Valigator.Text.Json/ValigatorJsonPropertyHandler.cs (hunk 7 above)
Comment (NathanMagnus) [reply Valigator.Text.Json/ValigatorJsonPropertyHandler.cs]: Sprouted out.

File: Valigator.Core/Valigator.Core.csproj (hunk 3 above)
Comment (NathanMagnus) [reply Valigator.Core/Valigator.Core.csproj:17]: This was removed automatically because it isn't needed.
